site/app/controllers/course/CourseMaterialsController.php

```diff
@@ -465,63 +497,96 @@ public function ajaxUploadCourseMaterialsFiles() {
                     else {
                         if (!@copy($uploaded_files[1]["tmp_name"][$j], $dst)) {
-                            return $this->core->getOutput()->renderResultMessage("ERROR: Failed to copy uploaded file {$uploaded_files[1]["name"][$j]} to current location.", false);
+                            return JsonResponse::getErrorResponse("Failed to copy uploaded file {$uploaded_files[1]['name'][$j]} to current location.");
                         }
                         else {
-                            if (!(is_null($sections))) {
-                                $sections_exploded = @explode(",", $sections);
-                                if ($sections_exploded == null) {
-                                    $sections_exploded = [];
-                                }
-                                $json[$dst] = ['release_datetime' => $release_time, 'sections' => $sections_exploded, 'hide_from_students' => $hide_from_students, 'external_link' => $is_external_link_file, 'sort_priority' => $sort_priority];
-                            }
-                            else {
-                                $json[$dst] = ['release_datetime' => $release_time, 'hide_from_students' => $hide_from_students, 'external_link' => $is_external_link_file, 'sort_priority' => $sort_priority];
-                            }
+                            $details['type'][$index] = $is_external_link_file ? CourseMaterial::LINK : CourseMaterial::FILE;
+                            $details['path'][$index] = $dst;
+                            $index++;
                         }
                     }
-                    //
                 }
                 else {
-                    return $this->core->getOutput()->renderResultMessage("ERROR: The tmp file '{$uploaded_files[1]['name'][$j]}' was not properly uploaded.", false);
+                    return JsonResponse::getErrorResponse("The tmp file '{$uploaded_files[1]['name'][$j]}' was not properly uploaded.");
                 }
                 // Is this really an error we should fail on?
                 if (!@unlink($uploaded_files[1]["tmp_name"][$j])) {
-                    return $this->core->getOutput()->renderResultMessage("ERROR: Failed to delete the uploaded file {$uploaded_files[1]["name"][$j]} from temporary storage.", false);
+                    return JsonResponse::getErrorResponse("Failed to delete the uploaded file {$uploaded_files[1]['name'][$j]} from temporary storage.");
+                }
+            }
+        }
+        if ($dirs_to_make != null) {
+            $i = -1;
+            $new_paths = [];
+            foreach ($dirs_to_make as $dir) {
+                $cm = $this->core->getCourseEntityManager()->getRepository(CourseMaterial::class)->findOneBy(
+                    ['path' => $dir]
+                );
+                if ($cm == null && !in_array($dir, $new_paths)) {
+                    $details['type'][$i] = CourseMaterial::DIR;
```

Comment:
Why are you making `DIR` type course materials here, but not when you convert from the JSON file to DB structure?

Reply:
In the migration these types do get created at the end:

https://github.com/Submitty/Submitty/blob/f2ce5e630a7f169d8a5406124d47461c160018fc/migration/migrator/migrations/course/20210610100802_course_materials_json_to_db.py#L121-L141